src/state_machine/coordinator/fire.rs

```diff
@@ -0,0 +1,1507 @@
+            }
+            State::NonceGather(is_taproot, merkle_root) => {
+                prev_state == &State::NonceRequest(*is_taproot, *merkle_root)
+                    || prev_state == &State::NonceGather(*is_taproot, *merkle_root)
+            }
+            State::SigShareRequest(is_taproot, merkle_root) => {
+                prev_state == &State::NonceGather(*is_taproot, *merkle_root)
+            }
+            State::SigShareGather(is_taproot, merkle_root) => {
+                prev_state == &State::SigShareRequest(*is_taproot, *merkle_root)
+                    || prev_state == &State::SigShareGather(*is_taproot, *merkle_root)
+            }
+        };
+        if accepted {
+            debug!("state change from {:?} to {:?}", prev_state, state);
+            Ok(())
+        } else {
+            Err(Error::BadStateChange(format!(
+                "{:?} to {:?}",
+                prev_state, state
+            )))
+        }
+    }
+}
+
+impl<Aggregator: AggregatorTrait> CoordinatorTrait for Coordinator<Aggregator> {
+    /// Create a new coordinator
+    fn new(config: Config) -> Self {
+        Self {
+            aggregator: Aggregator::new(config.num_keys, config.threshold),
+            config,
+            current_dkg_id: 0,
+            current_sign_id: 0,
+            current_sign_iter_id: 0,
+            dkg_public_shares: Default::default(),
+            party_polynomials: Default::default(),
+            public_nonces: Default::default(),
+            signature_shares: Default::default(),
+            aggregate_public_key: None,
+            signature: None,
+            schnorr_proof: None,
+            message: Default::default(),
+            dkg_wait_signer_ids: Default::default(),
+            nonce_recv_key_ids: Default::default(),
+            sign_wait_signer_ids: Default::default(),
+            sign_recv_key_ids: Default::default(),
+            state: State::Idle,
+            dkg_public_start: None,
+            dkg_end_start: None,
+            nonce_start: None,
+            sign_start: None,
+            malicious_signer_ids: Default::default(),
+        }
+    }
+
+    /// Retrieve the config
+    fn get_config(&self) -> Config {
+        self.config.clone()
+    }
+
+    /// Process inbound messages
+    fn process_inbound_messages(
+        &mut self,
+        packets: &[Packet],
+    ) -> Result<(Vec<Packet>, Vec<OperationResult>), Error> {
+        let mut outbound_packets = vec![];
+        let mut operation_results = vec![];
+        for packet in packets {
+            let (outbound_packet, operation_result) = self.process_message(packet)?;
+            if let Some(outbound_packet) = outbound_packet {
+                outbound_packets.push(outbound_packet);
+            }
+            if let Some(operation_result) = operation_result {
+                operation_results.push(operation_result);
+            }
+        }
+
+        let (outbound_packet, operation_result) = self.process_timeout()?;
```

Comment:
@jferrant I originally only processed timeouts if there were no packets received, but this seemed like a bad idea since clients could just send useless messages to prevent timeouts from being recognized.

Do you agree with this logic?  This is one of the things that's kept me from merging this.

Reply:
hm is a timeout meant to indicate a round timeout or a communication timeout? I think if its meant to be a "round" timeout. I.e. you took too damn long to run dkg or sign or whatever, then this makes sense.

Reply:
Yeah it's supposed to mean "X took longer than Y to happen"